dso-l1/src/main/java/com/tc/object/TransactionSource.java

```diff
@@ -19,32 +19,39 @@
 package com.tc.object;
 
 import com.tc.object.tx.TransactionID;
-
-import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.BitSet;
 import java.util.concurrent.atomic.AtomicLong;
 
 public class TransactionSource {
 
-  private static final AtomicLong nextTransactionId = new AtomicLong(1L);
-
-  private static final ConcurrentSkipListSet<Long> pending = new ConcurrentSkipListSet<>();
+  private final AtomicLong current = new AtomicLong();
+  private long retiredBase = 1L;
+  private int retirePosition = 0;
+  private BitSet retired = new BitSet();
 
   public TransactionID create() {
-    long txn;
-    while (!pending.add((txn = nextTransactionId.getAndIncrement())));
-    return new TransactionID(txn);
+    return new TransactionID(current.incrementAndGet());
+  }
+
+  public synchronized TransactionID oldest() {
+    retirePosition = retired.nextClearBit(retirePosition);
```

Comment:
Wouldn't it be easier to make 'oldest' a volatile and have an unsynced read here (prevent contention on message send), and then update it during the retire call?

Reply:
I'm not sure what the read guarantees are for BitSet.  I was erroring on the side of caution.

Reply:
I see what you are saying now